Adjust Cron when Custom Day Start is Changed Fixes #4488

carolstone opened the pull request:
The problem was that cron could be tricked into running over and over again by pushing the offset of dayStart forward an hour at a time. This solution will only allow cron to be run twice in quick succession if the user changes dayStart backward (the extreme case is allowing cron to run with a dayStart of 23 and then changing the dayStart offset to 0). But this behavior is much more logical than having cron run again if a player allows cron to run and then sets the offset to one hour in the future.

Modified common/script/index.coffee to display dayStart, but send back a variable of dayStart for processing, instead of having the object directly updated at the time of edit.

Modified website/views/options/settings.jade to allow a timestamp to be
cast to an iso timestamp because lastCron is saved as an iso timestamp,
and as a straight moment()-based timestamp in order to compare timestamps.

Modified website/public/js/controllers/settingsCtrl.js to do the
following:
   Accept the new proposed dayStart
   Calculate the timestamp of the old dayStart and the new dayStart

   These two steps were necessary because dayStart is an integer between
   and including 0 and 23. I tested and confirmed the current production
   dayStart does not allow 24 to be entered but does allow 0. I was
   careful to NOT change how that worked.

   Cast the old dayStart, the new dayStart and lastCron to the moment()
   version of time so they could be compared.

   Cast the new dayStart to the iso timestamp for storage in lastCron

   The important change is the following:

   if oldDayStart < lastCron AND lastCron < newDayStart then lastCron
   should be set to newDayStart. I modified this to include if oldDayStart
   = lastCron, although I think that's pretty unlikely to be possible.
   When I tested this, my lastCron was a mere 17 seconds after my
   oldDayStart, so it seemed to me that the equal case should be included
   with the less than case.

carolstone committed:
Modified common/script/index.coffee to display dayStart, but send back a
variable of dayStart for processing, instead of having the object directly
updated at the time of edit.

Modified website/views/options/settings.jade to allow a timestamp to be
cast to an iso timestamp because lastCron is saved as an iso timestamp,
and as a straight moment()-based timestamp in order to compare timestamps.

Modified website/public/js/controllers/settingsCtrl.js to do the
following:
   Accept the new proposed dayStart
   Calculate the timestamp of the old dayStart and the new dayStart

     These two steps were necessary because dayStart is an integer between
     and including 0 and 23. I tested and confirmed the current production
     dayStart does not allow 24 to be entered but does allow 0. I was
     careful to NOT change how that worked.

   Cast the old dayStart, the new dayStart and lastCron to the moment()
   version of time so they could be compared.

   Cast the new dayStart to the iso timestamp for storage in lastCron

   The important change is the following:

   if oldDayStart < lastCron AND lastCron < newDayStart then lastCron
   should be set to newDayStart. I modified this to include if oldDayStart
   = lastCron, although I think that's pretty unlikely to be possible.
   When I tested this, my lastCron was a mere 17 seconds after my
   oldDayStart, so it seemed to me that the equal case should be included
   with the less than case.

crookedneighbor commented on website/public/js/controllers/settingsCtrl.js:
```diff
@@ -66,12 +66,23 @@ habitrpg.controller('SettingsCtrl',
-    $scope.saveDayStart = function(){
-      var dayStart = +User.user.preferences.dayStart;
-      if (_.isNaN(dayStart) || dayStart < 0 || dayStart > 24) {
+    $scope.passDayStart = User.user.preferences.dayStart;
```
Can you explain what pass in passDayStart means? I keep thinking it's a misspelling of past, but am not sure.

carolstone replied:
If you look at my change to the jade module, you'll see I'm replacing a direct reference to user.preferences.dayStart with this variable and then passing it back as an argument to the function saveDayStart. I named it passDayStart to try to try to warn the reader this is a variable that's being passed around. I'd love to use a clearer name. 

In order to make it both appear as a default when loading the page in the first place and then be the value selected by the player when the code ran later, I discovered I had to send it back as an argument to the function, and give it a different name when it gets to the function.

crookedneighbor replied:
I get what it's doing, and there's no problem with assigning a different variable name for the saveDayStart argument, it's just not intuitive to me what the name `passDayStart` means. Your explanation makes sense to me, but I think we can come up with a clearer name for it.

carolstone replied:
Let me know if you think of something, and I'll chat with my cleverer half about it. I'd be happy to change it!

crookedneighbor commented on website/public/js/controllers/settingsCtrl.js:
```diff
@@ -66,12 +66,23 @@ habitrpg.controller('SettingsCtrl',
+      var getNewStart = Shared.startOfDay({ dayStart: dayStart});
+      var isoNewStart = Shared.isoTimestamp(getNewStart);
+
+      if (dayStart == undefined || _.isNaN(dayStart) || dayStart < 0 || dayStart > 24) {
```
Is the check here for undefined necessary? Wouldn't it be caught by _.isNAN since undefined is not a number?

carolstone replied:
I added that when I was testing out hijacking the return alert(window.env.t(<message>)); for debugging purposes. I was doing stuff like deleting the value completely, putting in the letter 'k', and occasionally I noticed that it was falling through this test and telling me the value I had passed was undefined, so I added it. When I get my vagrant box working again (it just informed me it's "inaccessible" this morning), I'll test this again and get back to you on what value fails the _isNaN test.

crookedneighbor replied:
Interesting. According to the [lodash documentation](https://lodash.com/docs#isNaN), `_.isNaN` returns false for `undefined`. In that case, I'd go with testing `!dayStart` instead of `dayStart === undefined || _.isNaN(dayStart)`. `NaN` is a falsy value, as is `undefined`, `null`, `false`, `""`, etc.

carolstone replied:
I learned how to fix my vagrant box, so I can get right on this!

carolstone replied:
Well, this is interesting and brings up more questions. Do we want to allow fractions? Is lodash not working on my machine for some reason? I'm really stumped now. Here is the result of my testing:

testing options for case to kick off message:
"please enter a number between 0 and 24"

if ( <result of test> || dayStart < 0 || dayStart > 24 ) {message}

Test 1: pass in the number zero, desired response = false
1. alert(dayStart); 0
2. alert(!dayStart); true
3. alert(isNaN(dayStart)); false
4. alert(!isNaN(dayStart)); true
5. alert(_.isNaN(dayStart)); false

success: 3,5

Test 2: pass in the number 1, desired response = false
1. alert(dayStart); 1
2. alert(!dayStart); false
3. alert(isNaN(dayStart)); false
4. alert(!isNaN(dayStart)); true
5. alert(_.isNaN(dayStart)); false

success: 2,3,5

Test 3: pass in the number 23, desired response = false
1. alert(dayStart);23
2. alert(!dayStart); false
3. alert(isNaN(dayStart)); false
4. alert(!isNaN(dayStart)); true
5. alert(_.isNaN(dayStart)); false

success: 2,3,5

Test 4: pass in the number -1, desired response = true
1. alert(dayStart); undefined
2. alert(!dayStart); true
3. alert(isNaN(dayStart)); true
4. alert(!isNaN(dayStart)); false
5. alert(_.isNaN(dayStart)); false

success: 2,3

Test 5: pass in the number 5.5, desired response ?
1. alert(dayStart); 5.5
2. alert(!dayStart); false
3. alert(isNaN(dayStart)); false 
4. alert(!isNaN(dayStart)); true
5. alert(_.isNaN(dayStart)); false

???

Test 6: pass in the number 23.5, desired response = ?
1. alert(dayStart); undefined
2. alert(!dayStart); true
3. alert(isNaN(dayStart)); true 
4. alert(!isNaN(dayStart)); false
5. alert(_.isNaN(dayStart)); false

???

Test 7: pass in the number 24, desired response = true
1. alert(dayStart); undefined
2. alert(!dayStart); true
3. alert(isNaN(dayStart)); true
4. alert(!isNaN(dayStart)); false
5. alert(_.isNaN(dayStart)); false

2,3

Test 8: pass in the letter k, desired response = true
1. alert(dayStart); undefined
2. alert(!dayStart); true
3. alert(isNaN(dayStart)); true
4. alert(!isNaN(dayStart)); false
5. alert(_.isNaN(dayStart)); false

2,3

Test 9: remove the value from the page, desired response = true
1. alert(dayStart); null
2. alert(!dayStart); true
3. alert(isNaN(dayStart)); false
4. alert(!isNaN(dayStart)); true
5. alert(_.isNaN(dayStart)); false

success: 2,4

crookedneighbor replied:
Not sure if this was a copy paste typo or not but, `if ( || dayStart < 0 || dayStart > 24 ) {message}` is missing `!dayStart`. It should be: `if (!dayStart || dayStart < 0 || dayStart > 24 ) {message}`

crookedneighbor replied:
Sorry, didn't read through your whole message before typing that. A little confused about what you're asking. 

As far as your questions about fractions, I do not believe we want to allow fractions or decimals. We should probably do `Math.floor` if those get passed in.

crookedneighbor replied:
Oh I see now. Yeah, checking for `!dayStart` by itself won't work because `!0` equals true, but 0 is a valid value.

In that case, I'd do what you did originally, but change the `==` to `===`

carolstone replied:
The only thing I don't like about `!dayStart` is it returns true for zero (Test case 1).

I did a bunch of tests. If I apply `Math.floor` right away in the var statement, `_.isNaN(dayStart)` works quite nicely, but the empty case, which was returning `null` in my previous tests, suddenly decided it was really a zero, and I don't think we want to just set someone's dayStart offset to zero if they happen leave the control after deleting the existing value. I was leaning toward `(!dayStart && dayStart != 0)`, but I also didn't like that `23.5` doesn't behave the same as `22.5` because the control passes anything greater than `23` as undefined.

On a lark, I tested `dayStart != Math.floor(dayStart)`, and I think that might be a good solution. It returns true for any values in `-1, 5.5, k, 23.5, 24, 25, null`. It returns false for all values in `0, 1, 5, 23`. This means we expect people to understand 5.5 is an invalid value when they get the error message, which really is implied by `_____:00 (24 hr clock)` on the page, but not so clear in the error message.

So if you agree, I'll use `if ( dayStart != Math.floor(dayStart) || dayStart < 0 || dayStart > 24 ) {error message}`.

crookedneighbor replied:
Sounds good to me.

crookedneighbor commented on website/public/js/controllers/settingsCtrl.js:
```diff
@@ -66,12 +66,23 @@ habitrpg.controller('SettingsCtrl',
         return alert(window.env.t('enterNumber'));
       }
+      if (Shared.momentTimestamp(getOldStart) <= Shared.momentTimestamp(lastCron) && Shared.momentTimestamp(lastCron) <  Shared.momentTimestamp(getNewStart)) {
+        User.set({ 'lastCron' : isoNewStart});
```
I'd like to pull all these lines regarding a new cron time out into a private method.

crookedneighbor replied:
Also, shared.startOfDay should return a moment object, so you shouldn't need to wrap those variables that are created with it in the moment time stamp function.

carolstone replied:
_facepalm_ Right. After casting it to something I could read, I forgot why I couldn't read the original format.

carolstone replied:
Should I put these lines into a private method in settingsCtrl.js, or should I push this back up to index.coffee in place of the items I exposed there? Or somewhere else?

crookedneighbor replied:
I would prefer a private method in the settingsCtrl.js file I can't think of a good reason to expose something this specific to anything outside of the settings controller.

crookedneighbor commented:
Really good start. I'm excited to see this get merged in.

crookedneighbor commented:
@carolstone Any update on this?

carolstone commented:
I received Alys' note and I hope I edited the title correctly. I'll stop holding this up. I'll check in the work I have over the week-end and propose an alternative solution after.

carolstone committed:
Renamed variables for clarity.

Moved calculation to actually change lastcron into its own function.

Changed validation errormessage to run: if ( dayStart != Math.floor(dayStart) || dayStart < 0 || dayStart > 24 ) {error message}.

Created new calculation for daystart because daystart could never be
future to the current timestamp. This "fix" has broken the whole endeavor,
though, because cron runs immedately when newlastcron > now and now >
oldlastcron.

crookedneighbor commented:
Sorry it's taken me so long to get back to you. Hoping to work on this today and see if we can get it working.

crookedneighbor commented:
@carolstone Your commits have been merged in. Thanks for starting this. 

This will count towards a contributor tier! What's your User ID? You can find it in [Settings > API](https://habitrpg.com/#/options/settings/api). Be sure to only copy the User ID and _not_ your API Token.

crookedneighbor commented:
Also, we're going to announce this change in a Bailey. What username would you like to be credited as? If we don't hear back from you, we'll list it as carolstone.

carolstone commented:
Yipes! Well that was a surprise. I'm sorry I fell off the face of the earth on you!  Work got really, really nasty busy and then I got sick and now I'm back to sick and working too much. I'm not entirely sure I trust that I really fixed this - I guess we're going to find out! My userID is fccb1fc7-a4fc-465b-9dcc-0cfdf37cbb59.

_edit by Alys Tue Nov  1 14:38:12 AEST 2016: Tier awarded._

burbel commented:
I believe that I just died to this item and wiped my party, as well.  Stayed up late last night, past 4am, and my cron ran and marked two tasks I was actually finishing (before bed chores) as missed.  I then moved my cron time forward two hours, to 6am, and went to bed (didn't mark any task).

Cron ran again after 6am and wiped the entire party.  Reset every streak I had made.

Alys commented:
@burbel If you change your Custom Day Start time on the website, you'll see a pop-up window that tells you when cron will next run. If that date and time doesn't suit you (e.g., it's too soon and you haven't finished your Dailies), then you can cancel out of that window and the CDS won't be changed. You can then make the CDS change later after finishing your Dailies. If you change your CDS on the iOS app, you don't get the same warning, but that will be added in future (https://github.com/HabitRPG/habitrpg-ios/issues/296). Until then, I recommend that you use the website for further CDS changes.

burbel commented:
Alys, I did change that on the website - while I have downloaded the application for iOS, I use it rarely if at all.  The change was made on the website, and there was no pop-up window that was displayed (Chrome) with information on the change.  It was doubly odd because having had the cron run for a single day at 4am, moving the chron to 6am re-ran the same day again (in both cases, the previous calendar day).  Are all times and dates tracked to UTC in HabitRPG?

Alys commented:
@burbel I've just tried changing my Custom Day Start time and the pop-up window did appear. Can you please try it again now (you do not have to confirm the change). If you do not see the pop-up window this time, please log that as a new bug. Note that the bug is that the confirmation window did not appear, not that cron ran again - there can be situations where it is appropriate for cron to run soon after changing the CDS and the pop-up window is designed to give you an appropriate warning about that. 

I'm not sure what you mean by "tracked to UTC". The Custom Day Start time is always in the timezone of the PC or mobile device(s) that you are using Habitica on.

burbel commented:
Alys, I've done a bit of testing (sorry for the delay) and I do get a popup notification changing times, including for the times that I originally tried to modify.  I do let Habitica run in a browser tab for days at a time - it's possible that one of the times I encountered an error had left things in a bad state.  

With the update explaining the behavior, what I experienced is definitely not a bug.  I will continue to check to see if I can find a solid reproduction case for the popup (overlay window, it seems) to not appear.  Thanks for taking the time to respond.
